pkg/watermark/fetch/edge_fetcher.go

```diff
@@ -71,25 +85,45 @@ func (e *edgeFetcher) GetWatermark(inputOffset isb.Offset) wmb.Watermark {
 	var epoch int64 = math.MaxInt64
 	var allProcessors = e.processorManager.GetAllProcessors()
 	for _, p := range allProcessors {
+		// headOffset is used to check whether this pod can be deleted.
+		headOffset := int64(-1)
+		// iterate over all the timelines of the processor and get the smallest watermark
 		debugString.WriteString(fmt.Sprintf("[Processor: %v] \n", p))
-		var t = p.GetOffsetTimeline().GetEventTime(inputOffset)
-		if t == -1 { // watermark cannot be computed, perhaps a new processing unit was added or offset fell off the timeline
-			epoch = t
-		} else if t < epoch {
-			epoch = t
+		for index, tl := range p.GetOffsetTimelines() {
+			// we only need to check the timelines of the partition we are reading from
+			if index == int(fromPartitionIdx) {
+				var t = tl.GetEventTime(inputOffset)
+				if t == -1 { // watermark cannot be computed, perhaps a new processing unit was added or offset fell off the timeline
+					epoch = t
+				} else if t < epoch {
+					epoch = t
+				}
+			}
+			// get the highest head offset among all the partitions of the processor so we can check later on whether the processor
+			// in question is stale (its head is far behind)
+			if tl.GetHeadOffset() > headOffset {
+				headOffset = tl.GetHeadOffset()
+			}
 		}
-		if p.IsDeleted() && (offset > p.GetOffsetTimeline().GetHeadOffset()) {
-			// if the pod is not active and the current offset is ahead of all offsets in Timeline
+
+		// if the pod is not active and the head offset of all the timelines is less than the input offset, delete the processor
+		// (this means we are processing data later than what the stale processor has processed)
+		if p.IsDeleted() && (offset > headOffset) {
 			e.processorManager.DeleteProcessor(p.GetEntity().GetName())
 		}
 	}
 	// if there are no processors
 	if epoch == math.MaxInt64 {
 		epoch = -1
 	}
+	// update the last processed watermark for the partition
+	e.lastProcessedWm[fromPartitionIdx] = epoch
+	// get the smallest watermark among all the partitions
+	// since we cannot compare the offset of different partitions, we get the smallest among the last processed watermarks of all the partitions
+	minEpoch := e.getMinFromLastProcessed(epoch)
 	e.log.Debugf("%s[%s] get watermark for offset %d: %+v", debugString.String(), e.bucketName, offset, epoch)
 
-	return wmb.Watermark(time.UnixMilli(epoch))
+	return wmb.Watermark(time.UnixMilli(minEpoch))
 }
```

Comment:
This is a bit of a nitpick, but I'm thinking about whether we should consider separating this into two methods. If I look at the method signature itself `GetWatermark(inputOffset isb.Offset, fromPartitionIdx int32)` it seems to suggest to me that we're getting the watermark just from this partition, but we are actually getting it from all of the partitions. I'm wondering what if we broke it into two functions: something like `ProcessPartitionOffset(inputOffset isb.Offset, fromPartitionIdx int32)` and `GetWatermark()`? What do you think?

Reply:
I would say

`GetCurrentPartitionWatermark(inputOffset isb.Offset, partitionIdx int32)`
and
`GetEdgeWatermark()`

and we can refactor this in the next PR

Reply:
I don't think we need separate methods.

Reply:
I still think the message signature's a little confusing and I'm a fan of making it more modular, but I'm okay if we do it at a later time, as it would be good to merge this now.

Reply:
> I still think the message signature's a little confusing and I'm a fan of making it more modular, but I'm okay if we do it at a later time, as it would be good to merge this now.

The signature is actually good IMO since it's a method of edge, edge.GetWatermark is the watermark for the edge which includes all the partitions

I just submitted a suggestion to add how we get the watermark in the godoc for more details.